framework/src/base/FEProblem.C

```diff
@@ -3459,10 +3459,14 @@ FEProblem::computeResidual(NonlinearImplicitSystem &/*sys*/, const NumericVector
   {
     computeResidualType(soln, residual, _kernel_type);
   }
-  catch(MooseException & e)
-  {
-    mooseError("Impossible");
```

Comment:
I put that "Impossible" error there because I'd knew sooner or later we'd hit it!

Take a look at `computeResidualType()`: 
https://github.com/idaholab/moose/blob/devel/framework/src/base/FEProblem.C#L3479-L3520

If you scan through that code you'll see we are computing several different kinds `execute_on = linear` objects: UserObjects, Transfers, AuxKernels, etc. Most of the parallel exception handling is being handled properly for those objects thanks to the magic of inheritance. (i.e. The logic for the MooseException is baked into ThreadedElementLoop and ThreadedNodalLoop). However it's unlikely as MOOSE evolves that's well manage to handle everything as the code changes. Case and point: Take a look now at `ComputeUserObjects()`:
https://github.com/idaholab/moose/blob/devel/framework/src/base/FEProblem.C#L2541-L2555

What happens if we throw an exception from a GeneralUserObject with `execute_on = linear`? There is no parallel exception handling anywhere in that code path so with this "fix" we'd swallow the exception and Move on - NOPE!

I think the design was that if we deal with an exception, we also swallow it at the same place. Therefore if an exception gets caught here, it's currently not properly handled. This fix will actually create really nasty bugs for somebody to find later!